### DISCUSSION TITLE
Confused on the purpose of TabIndexs List

### DISCUSSION THREAD

**Discussion by dasien** [tui-cs/Terminal.Gui#3319] [Q&A] [answered]
Hello,

I was hoping to get some clarity on the purpose of the `View.TabIndexes` list.  What i was trying to accomplish was to join two views together, and have the first control on View2 `TabIndex` be + 1 from the last control in `View1.TabIndexes`. 

What i am finding is that regardless of what i put for the `TabIndex` value for controls in View 1, even if i say that certain controls have no `TabIndex` and i set the `TabStop` to false, the views with no `TabIndex` and `TabStop` set to false are still listed in the `TabIndexes` list which inflates the list with controls.  As a consequence, the value i put in for `TabIndex` for controls which i want to interact with the `Tab` key is overridden by their order in the `TabIndexes` list.

Thus, if i have a view with a `Label` with `TabStop = false` and nothing set for `TabIndex`, and a `TextField` with `TabStop = true` and `TabIndex = 1`, and I repeat that pattern say 3 times incrementing the `TabIndex` by 1 for each `TextField` so the next `TextField.TabIndex = 2` then 3, etc. the values reported for the `TextField` in the parent `View.Tablndexes` list is 1,3,5,7 respectively rather than the 1,2,3,4 i set.

I am confused why something is overwriting my set `TabIndex` values on controls at all and if there is a way for me to turn that off.  I could understand the system taking a best guess if i don't put values in and don't also turn off the fact that a particular `View` should be a tab stop, but it is overwriting the values which are present and making it difficult for me to compute the correct `TabIndex` of a joined `View`.

**Comment by BDisp** [comment]
`Label` by default has `CanFocus` as false. Thus setting the `TabIndex` or `TabStop` has no influence. Also I think that is preferred set that properties after the views are added to their superviews and before run the application. If still it doesn't have the index you are setting please post a sample app that reproduce the issue, please.

**Comment by dasien** [comment]
Hi BDisp,

I did some testing last night with some sample code and the results were somewhat as you suggested, but in other ways strange.  I created three `Label` and three `TextField` views, and laid them out in three rows like so `Label` `Textfield`.  It seems like the engine goes from top left to lower right as it decides the `TabIndex` order of "tab-able" controls, is that right?  I only say that because my first run was without any intervention (i.e. not setting the `TabIndex` property at all resulted in the `TabIndexs` List sowing an element pattern of `Label` then `TextField`, 3 times.  The `Label` all had a TabIndex of -1, which i assume means that they are not to respond to Tab keys (because they don't take focus by default).  The `TextField` had 1,3,5 for their respective `TabIndex` properties, which makes sense given that i didn't pass any value in and they are the [1],[3], and [5] element in the `TabIndexes` `IList`.

For my second run, i added a value to the `TabIndex` for each `TextField` and then added them to the `View`.  Again, as you indicated, the values i added were overridden by whatever the layout engine determined to be appropriate, and the values remained 1,3,5 for the `TabIndex` of the `TextField` controls.

Where things get strange is when i add the `TextField` to the `View` then set the `TabIndex` property.  If i set those values to 0,1,2 for `TextField` 1, 2, and 3 respectively, The `TabIndexes` `List` shows the `TextField` in order in the first three elements of the `List` and the `Label` i the last three places. This has the unfortunate side effect of - in my use case - when i look at the `View.TabIndexes.Last()` which ostensibly be the last `TabIndex`, what is in that last position is a `Label` with a `TabIndex` of -1. 

Interestingly, if i set the values for `TabIndex` on the `TextField` to 3,4,5, after setting those controls to the `View` the resulting `TabIndexes` list shows `Label1` in element [0]  in the list,  then `TextField` 1 & 2 as the [1] and [2] elements, then two `Label` as [3] and [4] and finally `TextField3` as [5].  In this case, the underlying system chose to overwrite the values after they were set to the `View` of two of the three `TextFields`, and leave the third as it was set.

I will paste the code below with the last test values set.  A breakpoint on the line `IList < View > indexes = this.TabIndexes;` will show the order that the system puts controls.  

In the end, i am sure there is a predictable method to the way the system is ordering controls, but it seems rather arbitrary from the outside, and the `TabIndexes` list seems tightly coupled to the controls on a view - whether they can be tabbed to or not.  Having the [n]th element of a list sometimes equal the value of the `TabIndex` of the `View` it represents and sometimes not (in the case of `Label` and maybe other `Views` which are by default not focus-able in which case it is `-1`) is confusing and limits the value of the list to a client.

```
using Terminal.Gui;

namespace TabIndexTest 
{
    public class TabIndexView : Toplevel
    {
        private Label label1;
        private TextField textField1;
        private Label label2;
        private TextField textField2;
        private Label label3;
        private TextField textField3;
        public TabIndexView() 
        {
            
            InitializeComponent(true);
            
            // Show values for tabindex of textfields.
            IList < View > indexes = this.TabIndexes;
        }
        
        private void InitializeComponent(bool tabindexaftervew ) {
            textField3 = new TextField();
            label3 = new Label();
            textField2 = new TextField();
            label2 = new Label();
            textField1 = new TextField();
            label1 = new Label();
            Width = Dim.Fill(0);
            Height = Dim.Fill(0);
            X = 0;
            Y = 0;
            Visible = true;
            Modal = false;
            IsMdiContainer = false;
            TextAlignment = TextAlignment.Left;
            label1.Width = 4;
            label1.Height = 1;
            label1.X = 1;
            label1.Y = 1;
            label1.Visible = true;
            label1.Data = "label1";
            label1.Text = "Tabstop for Textfield 1";
            label1.TextAlignment = TextAlignment.Left;
            Add(label1);
            textField1.Width = 10;
            textField1.Height = 1;
            textField1.X = 55;
            textField1.Y = 1;
            textField1.Visible = true;
            textField1.Secret = false;
            textField1.Data = "textField1";
            textField1.Text = "";
            textField1.TextAlignment = TextAlignment.Left;

            if (tabindexaftervew)
            {
                Add(textField1);
                textField1.TabIndex = 3;
            }

            else
            {
                textField1.TabIndex = 3;
                Add(textField1);
            }
            
            label2.Width = 4;
            label2.Height = 1;
            label2.X = 1;
            label2.Y = 2;
            label2.Visible = true;
            label2.Data = "label2";
            label2.Text = "Tabstop for textfield 2";
            label2.TextAlignment = TextAlignment.Left;
            Add(label2);
            textField2.Width = 10;
            textField2.Height = 1;
            textField2.X = 55;
            textField2.Y = 2;
            textField2.Visible = true;
            textField2.Secret = false;
            textField2.Data = "textField2";
            textField2.Text = "";
            textField2.TextAlignment = TextAlignment.Left;
            
            if (tabindexaftervew)
            {
                Add(textField2);
                textField2.TabIndex = 4;
            }

            else
            {
                textField2.TabIndex = 4;
                Add(textField2);
            }
            
            label3.Width = 4;
            label3.Height = 1;
            label3.X = 1;
            label3.Y = 3;
            label3.Visible = true;
            label3.Data = "label3";
            label3.Text = "Tabstop for textfield 3";
            label3.TextAlignment = TextAlignment.Left;
            Add(label3);
            textField3.Width = 10;
            textField3.Height = 1;
            textField3.X = 55;
            textField3.Y = 3;
            textField3.Visible = true;
            textField3.Secret = false;
            textField3.Data = "textField3";
            textField3.Text = "";
            textField3.TextAlignment = TextAlignment.Left;
            
            if (tabindexaftervew)
            {
                Add(textField3);
                textField3.TabIndex = 5;
            }

            else
            {
                textField3.TabIndex = 5;
                Add(textField3);
            }

            Width = 43;
            Height = 12;
            X = 26;
            Y = 11;
            Visible = true;
            Data = "view";
            TextAlignment = TextAlignment.Left;
            
        }
    }
}
```

**Reply by tig** [reply]
Is this v1?

I think there's a bug in the Toplevel code that erroneously re-orders the subviews of a toplevel screwing up tabindex. I fixed this in v2 but forgot to do it in v1. I'm not at a computer right now but it would explain what you're seeing.

**Comment by dasien** [comment]
Hi - yes this is v1.  Sorry i should have indicated that.

**Comment by dasien** [comment]
Hi - i was just curious if this was something that could be fixed in v1, or if i should wait and move to v2.  It would certainly help to have the tab order be controllable.

**Reply by BDisp** [reply, marked as answer]
I found the bug and I'll submit a PR with a fix. But for your example work you must set the tabindex after all views are added to the superview and not by setting only after add one at a time. So after all the views are added then set the tabindex for them. In the v2 the bug also exist and I'll submit a PR for him too. Remember that the order index isn't continuous because all the tabindexes that have a value of -1 are ignored. So, it may have 1, 3, 5, etc... If you want only a ordered of all tabindexes you filter them, like `TabIndexes.Select(t => t > -1)`.

**Reply by BDisp** [reply]
Here is the fix https://github.com/gui-cs/Terminal.Gui/pull/3352.

**Reply by BDisp** [reply]
@dasien here is a proof about what I've said.

```cs
		[Fact]
		public void TabIndex_Invert_Order_Added_One_By_One_Does_Not_Do_What_Is_Expected ()
		{
			var r = new View ();
			var v1 = new View () { Id = "1", CanFocus = true };
			r.Add (v1);
			v1.TabIndex = 2;
			var v2 = new View () { Id = "2", CanFocus = true };
			r.Add (v2);
			v2.TabIndex = 1;
			var v3 = new View () { Id = "3", CanFocus = true };
			r.Add (v3);
			v3.TabIndex = 0;

			Assert.False (r.TabIndexes.IndexOf (v1) == 2);
			Assert.True (r.TabIndexes.IndexOf (v1) == 1);
			Assert.False (r.TabIndexes.IndexOf (v2) == 1);
			Assert.True (r.TabIndexes.IndexOf (v2) == 2);
			// Only the last is in the expected index
			Assert.True (r.TabIndexes.IndexOf (v3) == 0);

			Assert.True (r.Subviews.IndexOf (v1) == 0);
			Assert.True (r.Subviews.IndexOf (v2) == 1);
			Assert.True (r.Subviews.IndexOf (v3) == 2);
		}

```

**Reply by dasien** [reply]
Hi @BDisp.  Thank you for the update and the fix!  I am curious what the purpose of the `TabIndexes` list is.  For my purpose, i was trying to adjust the tab order of controls after a 'View' was added to a base view.  In other words, i have a 'View' with some common elements which will always show, and a "detail" view which i embed in the middle of the base `View` (from a visual perspective).  Because i don't know which of the detail views might be loaded (they have a different number of controls in each), and because i want the user to be able to tab through all the controls in order, i need to adjust the tab index of the "footer" controls in the base `View` after the detail view has been inserted.

What i had been doing was getting that `TabIndexes` list and just setting the tab index of the first footer control to `TabIndexes.Count` + 1, but that isn't really correct.  It sounds like i need to find the highest value in that List, not simply the count, because there are `Label` controls in there and others, which have -1 as their value.

I will only adjust the values after all controls have been loaded, as you instruct and see the effect.  Thanks again for the fix!

**Reply by BDisp** [reply]
> Hi @BDisp. Thank you for the update and the fix! I am curious what the purpose of the `TabIndexes` list is. For my purpose, i was trying to adjust the tab order of controls after a 'View' was added to a base view. In other words, i have a 'View' with some common elements which will always show, and a "detail" view which i embed in the middle of the base `View` (from a visual perspective). Because i don't know which of the detail views might be loaded (they have a different number of controls in each), and because i want the user to be able to tab through all the controls in order, i need to adjust the tab index of the "footer" controls in the base `View` after the detail view has been inserted.
> 
`TabIndexes` list is someway as the `SubViews` list, but order by the `TabIndex` property, which values equal to -1 are ignored. The `SubViews` list ares order by the added views for redraw in the Z-Order. When the `TabIndex` is set before the view is added it set the value to 0 because the `SuperView` is null. If you want the same order as views are being added you don't need to set the `TabIndex`.

> What i had been doing was getting that `TabIndexes` list and just setting the tab index of the first footer control to `TabIndexes.Count` + 1, but that isn't really correct. It sounds like i need to find the highest value in that List, not simply the count, because there are `Label` controls in there and others, which have -1 as their value.
> 
If you want to control the `TabIndexes` value yourself then you must set the value to `view.TabIndex = TabIndexes.IndexOf (view);` after the view is added to the superview, but the `Add` already do that automatically.

> I will only adjust the values after all controls have been loaded, as you instruct and see the effect. Thanks again for the fix!

Yes, but only if you need to do that. e.g. you may need to add a view2 first than the view1 to use Pos/Dim of the view2, but you want the `TabIndex` of the view1 bigger than the `TabIndex` of the view2. In this case you have to set the `TabIndex` value accordingly.
